### DISCUSSION TITLE
Don't depend on trait order for lifecycle hooks

### DISCUSSION THREAD

**Discussion by jimmyklein-actual** [livewire/livewire#8943] [3. Feature ideas]
Hello,

I work on a project with laravel 11 and Livewire3. 

We use hook on components with traits
```php
class MyComponent extends Component
{
    use BoundModelAutoSave;
    use ConvertEmptyStringsToNull;
    //...
}
```

We also have a rule on pint that is `ordered_traits` that sort alphabetically the trait in all our codebase.

The method `callTraitHook` in `SupportLifecycleHooks` class goes through all the traits as they are declared in the component.
```php
function callTraitHook($name, $params = [])
{
    foreach (class_uses_recursive($this->component) as $trait) {
       $method = $name.class_basename($trait);

        if (method_exists($this->component, $method)) {
            wrap($this->component)->$method(...$params);
        }
    }
}
```

But we need to have the hook in`ConvertEmptyStringsToNull` executed in first. 
So we renamed it `AAConvertEmptyStringsToNull` to be in the first place 😞 

I find it pretty dangerous to rely on traits order to the execution of code.

I suggest to give the possibility to declare the hooks class in a method in the component, for example `loadLifecycleHooks`

```php
public function loadLifecycleHooks(): array
{
     return [
         ConvertEmptyStringsToNull::class,
         BoundModelAutoSave::class,
        //...
    ];
}
```

and this method should be call from `SupportLifecycleHooks` class

For example, something like this :
```php
    function callCustomHook($name, $params = [])
    {
        if (method_exists($this->component, 'loadLifecycleHooks') {
             foreach ($this->component->loadLifecycleHooks() as $hook) {
                  app($hook)->$name($this->component, ...$params);
             }
        } 
    }
```

What is your point of view about this ?

**Comment by EdgarsJoja** [comment]
+1 for this.
Just run into a bug, because auto format had sorted traits alphabetically, which changed the execution order and broke logic. Now I'm not sure if I want to export such functionality to these traits at all, due to this behind-the-scenes behavior. This is going to be very easy to unknowingly break.
